subsys/bluetooth/shell/bt.c

```diff
@@ -193,65 +193,84 @@ static bool data_cb(struct bt_data *data, void *user_data)
+	for (int8_t i = len-1; i >= 0; i--) {
+		shell_fprintf(ctx_shell, SHELL_INFO, "%02x", data[i]);
+	}
+}
 
-	shell_fprintf(ctx_shell, SHELL_INFO, "%*sType 0x%02x: ",
-		      strlen(scan_response_label), "",
-		      type);
-	for (uint8_t i = 0U; i < (scan_data_len / min_value_len); i++) {
-		if (i > 0L) {
-			shell_fprintf(ctx_shell, SHELL_INFO, ", ");
-		}
+static void print_data_set(uint8_t set_value_len, bool single_value,
+			   const uint8_t *scan_data, uint8_t scan_data_len)
+{
+	uint8_t idx = 0;
 
-		shell_fprintf(ctx_shell, SHELL_INFO, "0x");
-		for (uint8_t j = 0U; j < min_value_len; j++) {
-			shell_fprintf(ctx_shell, SHELL_INFO, "%02x",
-				      scan_data[i * min_value_len + j]);
+	do {
+		if (idx > 0) {
+			shell_fprintf(ctx_shell, SHELL_INFO, ",");
 		}
-	}
-
-	shell_fprintf(ctx_shell, SHELL_INFO, "\n");
+		print_data_hex(&scan_data[idx], set_value_len);
+		idx += set_value_len;
+		if (single_value)
+			break;
+	} while (idx+set_value_len <= scan_data_len);
 }
 
 static bool data_verbose_cb(struct bt_data *data, void *user_data)
 {
+	shell_fprintf(ctx_shell, SHELL_INFO, "%*sType 0x%02x: ",
+		strlen(scan_response_label), "", data->type);
+
 	switch (data->type) {
 	case BT_DATA_UUID16_SOME:
 	case BT_DATA_UUID16_ALL:
 	case BT_DATA_SOLICIT16:
+		print_data_set(2, false, data->data, data->data_len);
+		break;
 	case BT_DATA_SVC_DATA16:
-		print_data_set(data->type, 2, data->data, data->data_len);
+		print_data_set(2, true, data->data, data->data_len);
+		shell_fprintf(ctx_shell, SHELL_INFO, ",");
+		print_data_set(1, false, data->data+2, data->data_len-2);
 		break;
 	case BT_DATA_UUID32_SOME:
 	case BT_DATA_UUID32_ALL:
+		print_data_set(4, false, data->data, data->data_len);
+		break;
 	case BT_DATA_SVC_DATA32:
-		print_data_set(data->type, 4, data->data, data->data_len);
+		print_data_set(4, true, data->data, data->data_len);
+		shell_fprintf(ctx_shell, SHELL_INFO, ",");
+		print_data_set(1, false, data->data+4, data->data_len-4);
 		break;
 	case BT_DATA_UUID128_SOME:
 	case BT_DATA_UUID128_ALL:
 	case BT_DATA_SOLICIT128:
+		print_data_set(16, false, data->data, data->data_len);
+		break;
 	case BT_DATA_SVC_DATA128:
-		print_data_set(data->type, 16, data->data, data->data_len);
+		print_data_set(16, true, data->data, data->data_len);
+		shell_fprintf(ctx_shell, SHELL_INFO, ",");
+		print_data_set(1, false, data->data+16, data->data_len-16);
 		break;
 	case BT_DATA_NAME_SHORTENED:
 	case BT_DATA_NAME_COMPLETE:
 	case BT_DATA_BROADCAST_NAME:
-		shell_info(ctx_shell, "%*sType 0x%02x: %.*s",
-			   strlen(scan_response_label), "",
-			   data->type,  data->data_len, data->data);
+		shell_fprintf(ctx_shell, SHELL_INFO, "%.*s", data->data_len, data->data);
 		break;
 	case BT_DATA_PUB_TARGET_ADDR:
 	case BT_DATA_RAND_TARGET_ADDR:
 	case BT_DATA_LE_BT_DEVICE_ADDRESS:
-		print_data_set(data->type, BT_ADDR_SIZE, data->data, data->data_len);
+		print_data_set(BT_ADDR_SIZE, false, data->data, data->data_len);
+		break;
```

Comment:
Ditto when printing addresses it is worth considering using `bt_addr_to_str`

Reply:
To avoid a new buffer allocation for the string, I'd rather do a fprintf variant that prints the address correctly.  Also, we'd need a separate string buffer that is then sent to the fprintf if we go through bt_addr_to_str